<h3>INDEXING DATA FRAMES</h3>

In [2]:
import pandas as pd

In [5]:
data=pd.read_csv("/Users/bestecetin/Desktop/machine_learning/courses/datascience/(5) Manipulating Data Frames with Pandas/pokemon.csv")

In [6]:
data=data.set_index("#") #index 1'den başlıyor.
data.head()

Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
#                                                                               
1      Bulbasaur  Grass  Poison  45      49       49       65       65     45   
2        Ivysaur  Grass  Poison  60      62       63       80       80     60   
3       Venusaur  Grass  Poison  80      82       83      100      100     80   
4  Mega Venusaur  Grass  Poison  80     100      123      122      120     80   
5     Charmander   Fire     NaN  39      52       43       60       50     65   

   Generation  Legendary  
#                         
1           1      False  
2           1      False  
3           1      False  
4           1      False  
5           1      False

In [7]:
data["HP"][1] 

45

yukarıdaki set_index'i yapmasaydık [0] yazardık , Bulbasaur'ın HP değerini getirdik.

In [8]:
data.HP[1]

45

In [9]:
data.loc[1,["HP"]]

HP    45
Name: 1, dtype: object

In [10]:
data[["HP","Attack"]]

HP  Attack
#              
1    45      49
2    60      62
3    80      82
4    80     100
5    39      52
..   ..     ...
796  50     100
797  50     160
798  80     110
799  80     160
800  80     110

[800 rows x 2 columns]

<h3>SLICING DATA FRAMES</h3>

Series ve Dataframes arasındaki farklar:

In [12]:
print(type(data["HP"]))    #series'de tek []
print(type(data[["HP"]]))  #dataframe'de [[ ]]

<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>


In [13]:
data.loc[1:10,"HP":"Defense"] #1'den 10'a kadar HP ve Defense dahil arasındaki degerlerle beraber getir.

HP  Attack  Defense
#                      
1   45      49       49
2   60      62       63
3   80      82       83
4   80     100      123
5   39      52       43
6   58      64       58
7   78      84       78
8   78     130      111
9   78     104       78
10  44      48       65

In [14]:
data.loc[10:1:-1,"HP":"Defense"] #10'dan 1'e kadar HP ve Defense dahil arasındaki degerlerle beraber getir. ters slicing

HP  Attack  Defense
#                      
10  44      48       65
9   78     104       78
8   78     130      111
7   78      84       78
6   58      64       58
5   39      52       43
4   80     100      123
3   80      82       83
2   60      62       63
1   45      49       49

In [15]:
#speed'den sonra sonuna kadar tüm columnları al.
data.loc[1:10,"Speed":]

Speed  Generation  Legendary
#                               
1      45           1      False
2      60           1      False
3      80           1      False
4      80           1      False
5      65           1      False
6      80           1      False
7     100           1      False
8     100           1      False
9     100           1      False
10     43           1      False

<h3>FILTERING DATA FRAMES</h3>

In [19]:
#combining filters
first_filter=data.HP > 150
second_filter=data.Speed > 35
data[first_filter & second_filter]

Name  Type 1 Type 2   HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
#                                                                              
122    Chansey  Normal    NaN  250       5        5       35      105     50   
262    Blissey  Normal    NaN  255      10       10       75      135     55   
352    Wailord   Water    NaN  170      90       45       90       45     60   
656  Alomomola   Water    NaN  165      75       80       40       45     65   

     Generation  Legendary  
#                           
122           1      False  
262           2      False  
352           3      False  
656           5      False

In [21]:
data[data.Speed<15]

Name  Type 1 Type 2   HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
#                                                                              
231    Shuckle     Bug   Rock   20      10      230       10      230      5   
360   Trapinch  Ground    NaN   45     100       45       45       45     10   
487     Bonsly    Rock    NaN   50      80       95       10       45     10   
496   Munchlax  Normal    NaN  135      85       40       40       85      5   
659  Ferroseed   Grass  Steel   44      50       91       24       86     10   

     Generation  Legendary  
#                           
231           2      False  
360           3      False  
487           4      False  
496           4      False  
659           5      False

In [20]:
data.HP[data.Speed<15] #hıza göre filtrelediğin datanın HP'sini göster.

#
231     20
360     45
487     50
496    135
659     44
Name: HP, dtype: int64

<h3>TRANSFORMING DATA</h3>

In [22]:
#plain python functions:
def div(n):
    return n/2
data.HP.apply(div) 

#
1      22.5
2      30.0
3      40.0
4      40.0
5      19.5
       ... 
796    25.0
797    25.0
798    40.0
799    40.0
800    40.0
Name: HP, Length: 800, dtype: float64

data'nın HP'sine  div fonksiyonunu uyguladık <strong>(apply)</strong> ile.

aynısını lambda ile yapmak icin:

In [23]:
data.HP.apply(lambda n:n/2)

#
1      22.5
2      30.0
3      40.0
4      40.0
5      19.5
       ... 
796    25.0
797    25.0
798    40.0
799    40.0
800    40.0
Name: HP, Length: 800, dtype: float64

yeni bir column yapmak icin:

In [25]:
data["total_power"]=data.Attack+data.Defense
data.head()

Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
#                                                                               
1      Bulbasaur  Grass  Poison  45      49       49       65       65     45   
2        Ivysaur  Grass  Poison  60      62       63       80       80     60   
3       Venusaur  Grass  Poison  80      82       83      100      100     80   
4  Mega Venusaur  Grass  Poison  80     100      123      122      120     80   
5     Charmander   Fire     NaN  39      52       43       60       50     65   

   Generation  Legendary  total_power  
#                                      
1           1      False           98  
2           1      False          125  
3           1      False          165  
4           1      False          223  
5           1      False           95

<h3>INDEX OBJECTS AND LABELED DATA</h3>

In [26]:
print(data.index.name)

#


datanın index'inin adı # ve bunu değiştirmek için:

In [27]:
data.index.name="index_name"
data.head()

Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
index_name                                                               
1               Bulbasaur  Grass  Poison  45      49       49       65   
2                 Ivysaur  Grass  Poison  60      62       63       80   
3                Venusaur  Grass  Poison  80      82       83      100   
4           Mega Venusaur  Grass  Poison  80     100      123      122   
5              Charmander   Fire     NaN  39      52       43       60   

            Sp. Def  Speed  Generation  Legendary  total_power  
index_name                                                      
1                65     45           1      False           98  
2                80     60           1      False          125  
3               100     80           1      False          165  
4               120     80           1      False          223  
5                50     65           1      False           95

In [28]:
data1=data.copy()
#data'nın index'inin aralığını değiştirmek için:
data1.index=range(100,900,1) #100'den 900'e birer birer artsın.
data1.head()

Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  Sp. Def  \
100      Bulbasaur  Grass  Poison  45      49       49       65       65   
101        Ivysaur  Grass  Poison  60      62       63       80       80   
102       Venusaur  Grass  Poison  80      82       83      100      100   
103  Mega Venusaur  Grass  Poison  80     100      123      122      120   
104     Charmander   Fire     NaN  39      52       43       60       50   

     Speed  Generation  Legendary  total_power  
100     45           1      False           98  
101     60           1      False          125  
102     80           1      False          165  
103     80           1      False          223  
104     65           1      False           95

bir feature'ı index olarak tanımlayabiliriz: <br>
data=data.set_index("#") ya da: <br>
data.index=data["#"] <br>
ikisi de aynı şey

<h3>HIERARCHICAL INDEXING</h3> <br>
<li>Setting indexing</li>

In [36]:
data1=pd.read_csv("/Users/bestecetin/Desktop/machine_learning/courses/datascience/(5) Manipulating Data Frames with Pandas/pokemon.csv")

In [ ]:
data1=data.set_index(["Type 1","Type 2"])

data'nın indexini Type 1 ve Type 2 olarak set ediyorum. <br>
ilk index: <strong>outer</strong>, ikinci index ise: <strong>inner</strong> index oluyor.

In [43]:
data1.head(50)

#              Name   HP  Attack  Defense  Sp. Atk  Sp. Def  \
Type 1   Type 2                                                                 
Grass    Poison   1         Bulbasaur   45      49       49       65       65   
         Poison   2           Ivysaur   60      62       63       80       80   
         Poison   3          Venusaur   80      82       83      100      100   
         Poison   4     Mega Venusaur   80     100      123      122      120   
Fire     NaN      5        Charmander   39      52       43       60       50   
         NaN      6        Charmeleon   58      64       58       80       65   
         Flying   7         Charizard   78      84       78      109       85   
         Dragon   8  Mega Charizard X   78     130      111      130       85   
         Flying   9  Mega Charizard Y   78     104       78      159      115   
Water    NaN     10          Squirtle   44      48       65       50       64   
         NaN     11         Wartortle   59      63       80       65       80   
         NaN     12         Blastoise   79      83      100       85      105   
         NaN     13    Mega Blastoise   79     103      120      135      115   
Bug      NaN     14          Caterpie   45      30       35       20       20   
         NaN     15           Metapod   50      20       55       25       25   
         Flying  16        Butterfree   60      45       50       90       80   
         Poison  17            Weedle   40      35       30       20       20   
         Poison  18            Kakuna   45      25       50       25       25   
         Poison  19          Beedrill   65      90       40       45       80   
         Poison  20     Mega Beedrill   65     150       40       15       80   
Normal   Flying  21            Pidgey   40      45       40       35       35   
         Flying  22         Pidgeotto   63      60       55       50       50   
         Flying  23           Pidgeot   83      80       75       70       70   
         Flying  24      Mega Pidgeot   83      80       80      135       80   
         NaN     25           Rattata   30      56       35       25       35   
         NaN     26          Raticate   55      81       60       50       70   
         Flying  27           Spearow   40      60       30       31       31   
         Flying  28            Fearow   65      90       65       61       61   
Poison   NaN     29             Ekans   35      60       44       40       54   
         NaN     30             Arbok   60      85       69       65       79   
Electric NaN     31           Pikachu   35      55       40       50       50   
         NaN     32            Raichu   60      90       55       90       80   
Ground   NaN     33         Sandshrew   50      75       85       20       30   
         NaN     34         Sandslash   75     100      110       45       55   
Poison   NaN     35          Nidoran♀   55      47       52       40       40   
         NaN     36          Nidorina   70      62       67       55       55   
         Ground  37         Nidoqueen   90      92       87       75       85   
         NaN     38          Nidoran♂   46      57       40       40       40   
         NaN     39          Nidorino   61      72       57       55       55   
         Ground  40          Nidoking   81     102       77       85       75   
Fairy    NaN     41          Clefairy   70      45       48       60       65   
         NaN     42          Clefable   95      70       73       95       90   
Fire     NaN     43            Vulpix   38      41       40       50       65   
         NaN     44         Ninetales   73      76       75       81      100   
Normal   Fairy   45        Jigglypuff  115      45       20       45       25   
         Fairy   46        Wigglytuff  140      70       45       85       50   
Poison   Flying  47             Zubat   40      45       35       30       40   
         Flying  48            Golbat   75      8

Gördüğümüz gibi Type 1 ve Type 2 eskiden feature'du şuanda index oldular,<br>
Grass'ların hepsi Poison'muş, <br>
Water pokemonların Type2'si yokmuş

<h3>PIVOTING DATA FRAMES</h3> <br>
<li>pivoting: reshape tool</li>

In [45]:
dic={"treatment":["A","A","B","B"],"gender":["F","M","F","M"],"response":[10,45,5,9],"age":[15,4,72,65]}
df=pd.DataFrame(dic)
df

treatment gender  response  age
0         A      F        10   15
1         A      M        45    4
2         B      F         5   72
3         B      M         9   65

bir data frame oluşturduk feature olarak tedavi, cinsiyet, tedaviye cevap ve yaş ekledik.

In [46]:
#pivoting
df.pivot(index="treatment",columns="gender",values="response")

gender      F   M
treatment        
A          10  45
B           5   9

<strong>pivot</strong> ile index'i treatment, feature'ları gender verileri ise response yaptık.

<h3>STACKING and UNSTACKING DATAFRAME</h3> <br>
birden fazla indexe sahip data frame'leri unstack yapabiliyoruz.(indexlerinden bir tanesinde kurtulabiliyoruz.) <br>
level: unstack yapılacak index <br>
swaplevel: change inner and outer level index position

In [48]:
df1=df.set_index(["treatment","gender"]) 
df1

response  age
treatment gender               
A         F             10   15
          M             45    4
B         F              5   72
          M              9   65

veri setimdeki outer index:treatment, inner index: gender olsun dedik. 

In [49]:
df1.unstack(level=0) #treatment artık index değil

response    age    
treatment        A  B   A   B
gender                       
F               10  5  15  72
M               45  9   4  65

In [52]:
df1.unstack(level=1)  #gender index değil.

response     age    
gender           F   M   F   M
treatment                     
A               10  45  15   4
B                5   9  72  65

indexlerin yerlerini değiştirmek için yani inner ve outer'ın yerini:

In [53]:
df2=df1.swaplevel(0,1)
df2

response  age
gender treatment               
F      A                10   15
M      A                45    4
F      B                 5   72
M      B                 9   65

artık veri setimdeki outer index:gender, inner index: treatment oldu.

<h3>MELTING DATA FRAMES</h3> <br>

<li>Reverse of pivoting</li>

In [57]:
pd.melt(df,id_vars="treatment",value_vars=["age","response"])

treatment  variable  value
0         A       age     15
1         A       age      4
2         B       age     72
3         B       age     65
4         A  response     10
5         A  response     45
6         B  response      5
7         B  response      9

<strong>id_vars</strong> : treatment yani treatment sabit kalsın, <br>
<strong>value_vars</strong>:variable age ve response olsun value ise bunların tablodaki değerleri olsun.


<h3>CATEGORICALS AND GROUPBY</h3>

In [59]:
df

treatment gender  response  age
0         A      F        10   15
1         A      M        45    4
2         B      F         5   72
3         B      M         9   65

In [61]:
df.groupby("treatment").mean() #treatment'a göre grupla ve bunun ortalamasını al.

response   age
treatment                
A              27.5   9.5
B               7.0  68.5

In [64]:
df.groupby("treatment").age.mean()

treatment
A     9.5
B    68.5
Name: age, dtype: float64


treatment'e göre grupla sonra bunların yaşını al ve yaşların ortalamasını hesapla <br>
15+4/2 ve<br> 72+65/2 yani <br> 9.5 ve 68.5 yapar.

In [65]:
df.groupby("treatment").age.max() #treatment'e göre grupla sonra yaşların maximum olanını göster.

treatment
A    15
B    72
Name: age, dtype: int64

In [67]:
df.groupby("treatment")[["age","response"]].min()

age  response
treatment               
A            4        10
B           65         5

treatment'a göre grupla ve age ve response için minimumları bul